specification/batch/data-plane/Microsoft.Batch/stable/2018-03-01.6.1/BatchService.json

```diff
@@ -11513,7 +11513,7 @@
           "type": "integer",
           "format": "int32",
           "title": "The maximum number of times the task may be retried. The Batch service retries a task if its exit code is nonzero.",
-          "description": "Note that this value specifically controls the number of retries. The Batch service will try the task once, and may then retry up to this limit. For example, if the maximum retry count is 3, Batch tries the task up to 4 times (one initial try and 3 retries). If the maximum retry count is 0, the Batch service does not retry the task. If the maximum retry count is -1, the Batch service retries the task without limit."
+          "description": "Note that this value specifically controls the number of retries. The Batch service will try the task once, and may then retry up to this limit. For example, if the maximum retry count is 3, Batch tries the task up to 4 times (one initial try and 3 retries). If the maximum retry count is 0, the Batch service does not retry the task. If the maximum retry count is -1, the Batch service retries the task without limit. ResourceFiles and AppplicationPackages are only downloaded again if the task is retried on a new compute node. Batch will also retry tasks when a recovery operation is triggered on a compute node. Examples of recovery operations triggers include (but are not limited to) when an unhealthy compute node is rebooted or a compute node disappeared due to host failure. Retries due to recovery operation are independent of and are not counted against maxTaskRetryCount. All tasks should be idempotent, it means they need to tolerate being interrupted and restarted, without causing any corruption or duplicate data. Best practices recommended for long running tasks is to use checkpointing."
```

Comment:
"All tasks should be idempotent, it means..." replace by "All tasks should be idempotent.  This means tasks need to tolerate being... etc".  

I don't think the email on this text and/or placement has reached a conclusion yet.

Reply:
My attempt:
"Note that this value specifically controls the number of retries for the task executable due to nonzero exit code. The Batch service will try the task once, and may then retry up to this limit. For example, if the maximum retry count is 3, Batch tries the task up to 4 times (one initial try and 3 retries). If the maximum retry count is 0, the Batch service does not retry the task after the first attempt. If the maximum retry count is -1, the Batch service retries the task without limit. Resource files and application packages are only downloaded again if the task is retried on a new compute node. Batch will also retry tasks when a recovery operation is triggered on a compute node. Examples of recovery operations include (but are not limited to) when an unhealthy compute node is rebooted or a compute node disappeared due to host failure. Retries due to recovery operations are independent of and are not counted against the maxTaskRetryCount. Even if the maxTaskRetryCount is 0, an internal retry due to a recovery operation may occur. Because of this, all tasks should be idempotent. This means tasks need to tolerate being interrupted and restarted without causing any corruption or duplicate data. Best practices recommended for long running tasks is to use checkpointing." 

I think some/all of this text (specifically this bit: `Batch will also retry tasks when a recovery operation is triggered on a compute node. Examples of recovery operations include (but are not limited to) when an unhealthy compute node is rebooted or a compute node disappeared due to host failure. Retries due to recovery operations are independent of and are not counted against the maxTaskRetryCount. Even if the maxTaskRetryCount is 0, an internal retry due to a recovery operation may occur. Because of this, all tasks should be idempotent. This means tasks need to tolerate being interrupted and restarted without causing any corruption or duplicate data. Best practices recommended for long running tasks is to use checkpointing.` which is about internal retries should be moved to the root of the task object rather than buried down here on maxtaskRetryCount which is actually about _user_ retries not system/internal ones.

Reply:
Are we able to include new lines in the description documentation? The internal retries topic is really a separate concept than the user retries and it would be helpful to make a clear distinction when we branch off into it to prevent confusion.

I agree with Matt's point that idempotent tasks should be addressed  closer to the root as well because not everyone will care enough to read into retry policy. Generally when reading documentation people will start at the base and then delve deeper into the options as needed.